Added README.md for demo_nodes_py (#600)
ros2/demos: demo_nodes_py/README.md

## **What Is This?**

This demo ...

> This package consists of ROS 2 nodes:

Reviewer: Can remove "This demo ...",  and change the next line to,
"This package contains the following ROS 2 nodes:"

Author: Thank you for the review. Will follow up with the nits resolved shortly.

Author: Done.

1. `listener`
2. `talker`
3. `add_two_ints_client`
4. `add_two_ints_server`
5. `set_parameters_callback`
6. `add_two_ints_client_async`
7. `listener_qos`
8. `talker_qos`
9. `listerner_serialized`
10. `async_param_client`

## **Build**

Run this command to build the package:

```bash
colcon build --packages-select demo_nodes_py
```

## **Run**

### Basic Listener & Talker

> This runs `talker` and `listener` ROS 2 nodes which exchanges the following string with incremeting integer:

Reviewer: "...which exchange the following..." , remove the extra 's'

Author: Done.

> Hello World: <count_>

```bash
# Open new terminal
ros2 run demo_nodes_py talker
```

```bash
# Open new terminal
ros2 run demo_nodes_py listener
```

![](img/talker_listener.png)

### Server & Client

> This runs `add_two_ints_client` and `add_two_ints_server` ROS 2 client and server where  Server processes two integers sent from Client and publishes its sum to Client to be printed out.

Reviewer: Can remove the capitalization of "Server" and "Client". Also, "... where the server processes..." (add 'the')

Author: Done. Added 'the' in front of 'client' as well.

Changed to "... publishes its sum to the client" and "sent from the client".

#### Server

```bash
# Open new terminal
ros2 run demo_nodes_py add_two_ints_server
```

#### Client [Synchronous]

```bash
# Open new terminal
ros2 run demo_nodes_py add_two_ints_client
```

#### Client [Asynchronous]

```bash
# Open new terminal
ros2 run demo_nodes_py add_two_ints_client_async
```
![](img/server_client.png)

### QoS Listener & Talker

> Similar to previous, this runs `talker_qos` and `listener_qos` ROS 2 nodes which exchanges using the Quality of Service (QoS) Best Effort policy.

Reviewer: "... which exchange messages using the Quality of Service (QoS) best effort policy."

Author: Done.

```bash
# Open new terminal
ros2 run demo_nodes_cpp parameter_blackboard
```

> For more details on parameter_blackboard, please refer to demo_nodes_cpp README.md.

```bash
# Open new terminal
ros2 run demo_nodes_py async_param_client
```

### Serialized Subscriber

This runs `serialized_subscriber` ROS 2 node that subscribes to the ROS 2 topic, `/chatter`, and deserializes as well as prints out received strings:

```bash
# Open new terminal
ros2 run demo_nodes_py listener_serialized
```

![](img/serialized_subscriber.png)

## **Verify**

### Basic Listener & Talker

When executed correctly, strings should be printed to terminal similar to what is shown below:

```bash
# In terminal running talker
[INFO] [1674573022.635075580] [talker]: Publishing: "Hello World: 0"
[INFO] [1674573023.593075728] [talker]: Publishing: "Hello World: 1"
[INFO] [1674573024.592438479] [talker]: Publishing: "Hello World: 2"
```

```bash
# In terminal running listener
[INFO] [1674573025.634810166] [listener]: I heard: [Hello World: 0]
[INFO] [1674573026.596392653] [listener]: I heard: [Hello World: 1]
[INFO] [1674573027.596400384] [listener]: I heard: [Hello World: 2]
#...
```

### Basic Server & Client

> When executed correctly, strings should be printed to terminal similar to what is shown below:

Reviewer: "...printed to the terminal ..."

Author: Done.

#### Server
```bash
# In terminal running add_two_ints_server
[INFO] [1674553268.912391774] [add_two_ints_server]: Incoming request
a: 2 b: 3
```

#### Client [Synchronous]
```bash
# In terminal running add_two_ints_client
[INFO] [1674553268.912602310] [add_two_ints_client]: Result of add_two_ints: 5
```

#### Client [Asynchronous]
```bash
# In terminal running add_two_ints_client_async
[INFO] [1674553718.598690033] [add_two_ints_client]: Result of add_two_ints: 5
```

### QoS Listener & Talker

> When executed correctly, strings should be printed to terminal similar to what is shown below:

Reviewer: "...printed to the terminal ...."

Author: Done.

[INFO] [1674572633.021197421] [talker_qos]: Publishing: "Hello World: 2"
#...
```

```bash
# In terminal running listener_qos
[INFO] [1674572639.695289182] [listener_qos]: Best effort listener
[INFO] [1674572640.023684925] [listener_qos]: I heard: [Hello World: 0]
[INFO] [1674572641.023416691] [listener_qos]: I heard: [Hello World: 1]
[INFO] [1674572642.024505343] [listener_qos]: I heard: [Hello World: 2]
#...
```

### Set Parameters Callback
#### [Before]
Run `ros2 param get /set_parameters_callback param1` should print the following to terminal:
```bash
Double value is: 0.0
```
Run `ros2 param get /set_parameters_callback param2` should print the following to terminal:
```bash
Double value is 0.0
```
#### [Change]
Run `ros2 param set /set_parameters_callback param1 28.0` should print the following to terminal:
```bash
Set parameter successful
```
#### [After]
Run `ros2 param get /set_parameters_callback param1` should print the following to terminal:
```bash
Double value is: 28.0
```
Run `ros2 param get /set_parameters_callback param2` should print the following to terminal:
```bash
Double value is 4.0
```

### Asynchronous Parameter Client

> Running `ros2 run demo_nodes_cpp parameter_blackboard` should print output to the terminal similar to what is shown below:

Reviewer: "... print an output ....."

Author: Done.

## **What Is This?**

This demo ...

This package consists of ROS 2 nodes:

1. `listener`
2. `talker`
3. `add_two_ints_client`
4. `add_two_ints_server`
5. `set_parameters_callback`
6. `add_two_ints_client_async`
7. `listener_qos`
8. `talker_qos`
9. `listerner_serialized`
10. `async_param_client`

## **Build**

Run this command to build the package:

```bash
colcon build --packages-select demo_nodes_py
```

## **Run**

### Basic Listener & Talker

This runs `talker` and `listener` ROS 2 nodes which exchanges the following string with incremeting integer:

> Hello World: <count_>

```bash
# Open new terminal
ros2 run demo_nodes_py talker
```

```bash
# Open new terminal
ros2 run demo_nodes_py listener
```

![](img/talker_listener.png)

### Server & Client

> This runs `add_two_ints_client` and `add_two_ints_server` ROS 2 client and server which Server processes two integers sent from Client and publishes its sum to Client to be printed out.

Author: Done.

[INFO] [1674572633.021197421] [talker_qos]: Publishing: "Hello World: 2"
#...
```

```bash
# In terminal running listener_qos
[INFO] [1674572639.695289182] [listener_qos]: Best effort listener
[INFO] [1674572640.023684925] [listener_qos]: I heard: [Hello World: 0]
[INFO] [1674572641.023416691] [listener_qos]: I heard: [Hello World: 1]
[INFO] [1674572642.024505343] [listener_qos]: I heard: [Hello World: 2]
#...
```

### Set Parameters Callback
#### [Before]
Run `ros2 param get /set_parameters_callback param1` should print the following to terminal:
```bash
Double value is: 0.0
```
Run `ros2 param get /set_parameters_callback param2` should print the following to terminal:
```bash
Double value is 0.0
```
#### [Change]
Run `ros2 param set /set_parameters_callback param1 28.0` should print the following to terminal:
```bash
Set parameter successful
```
#### [After]
Run `ros2 param get /set_parameters_callback param1` should print the following to terminal:
```bash
Double value is: 28.0
```
Run `ros2 param get /set_parameters_callback param2` should print the following to terminal:
```bash
Double value is 4.0
```

### Asynchronous Parameter Client

> Running `ros2 run demo_nodes_cpp parameter_blackboard` should have similar strings printed to terminal similar to what is shown below:

Author: Done.

[INFO] [1674574042.989605537] [async_param_client]:         successful: True
[INFO] [1674574042.989847066] [async_param_client]:         value: One fish, two fish, Red fish, blue fish
[INFO] [1674574042.990086215] [async_param_client]:     bool_parameter:
[INFO] [1674574042.990325361] [async_param_client]:         successful: True
[INFO] [1674574042.990565005] [async_param_client]:         value: False
[INFO] [1674574042.990804701] [async_param_client]: Deleting parameters:
[INFO] [1674574042.991497679] [async_param_client]:     other_int_parameter:
[INFO] [1674574042.991746545] [async_param_client]:         successful: True
[INFO] [1674574042.991986980] [async_param_client]:         reason:
[INFO] [1674574042.992226951] [async_param_client]:     other_string_parameter:
[INFO] [1674574042.992465274] [async_param_client]:         successful: True
[INFO] [1674574042.992703259] [async_param_client]:         reason:
[INFO] [1674574042.992941566] [async_param_client]:     string_parameter:
[INFO] [1674574042.993179508] [async_param_client]:         successful: True
[INFO] [1674574042.993416599] [async_param_client]:         reason:
```

### Serialized Subscriber

> Running `ros2 run demo_nodes_py talker` should have string printed to termina in a similar manner: 

Reviewer 2: ```suggestion
Running `ros2 run demo_nodes_py talker` should print output to the terminal similar to the following: 
```

Author: Done.
